driver-sync/src/main/com/mongodb/client/internal/ClientSessionImpl.java

```diff
@@ -153,6 +157,49 @@ public void abortTransaction() {
         }
     }
 
+    @Override
+    public <T> T withTransaction(final TransactionBody<T> transactionBody) {
+        return withTransaction(TransactionOptions.builder().build(), transactionBody);
+    }
+
+    @Override
+    public <T> T withTransaction(final TransactionOptions options, final TransactionBody<T> transactionBody) {
+        outer:
+        while (true) {
+            T retVal;
+            try {
+                startTransaction(options);
+                retVal = transactionBody.execute();
+            } catch (RuntimeException e) {
+                if (transactionState == TransactionState.IN) {
+                    abortTransaction();
+                }
+                if (e instanceof MongoException) {
+                    if (((MongoException) e).hasErrorLabel(TRANSIENT_TRANSACTION_ERROR_LABEL)) {
+                        continue;
+                    }
+                }
+                throw e;
+            }
+            if (transactionState == TransactionState.IN) {
+                while (true) {
+                    try {
+                        commitTransaction();
+                        break;
+                    } catch (MongoException e) {
+                        if (e.hasErrorLabel(UNKNOWN_TRANSACTION_COMMIT_RESULT_LABEL)) {
```

Comment:
This isn't quite according to spec now, as there's another condition about write concern timeouts.  Deferring implementation though until there's a test for it.

Reply:
There will be a follow-on ticket for the final implementation.